qiskit/backends/_basebackend.py

```diff
@@ -0,0 +1,40 @@
+"""This module implements the abstract base class for backend modules.
+
+To create add-on backend modules subclass the Backend class in this module. 
+Doing so requires that the required backend interface is implemented.
+"""
+from abc import ABC, abstractmethod
+
+class BaseBackend(ABC):
+    
+    @abstractmethod
+    def __init__(self, qobj):
+        """initialize
+
+        This method should initialize the module and raise a FileNotFoundError 
+        exception if a component of the moduel is not available. 
+
+        Args:
+            qobj (dict): qobj dictionary
+        
+        Raises:
+            FileNotFoundError if backend executable is not available.
+        """
+        self._qobj = qobj
+        self._configuration = None # IMPLEMENT for your backend
```

Comment:
I think the nice thing about the discovery is that the backend developer does not have to put `import` statements in any qiskit file for the backend to be used. They only need to know the name of the backend they put into the configuration of their own backend module. Even for the number of backends we have this keeps the core code cleaner. 

I think you could be right about the user putting the module in the backends folder if they've installed with pip. Perhaps if they are developing backends they are more likely to use a git repo and would be comfortable with putting modules there. Still, maybe it's strange to have "local" modules mixed with qiskit ones. Perhaps we can have a ".qiskit/backends" folder in their home directory, or other locations, which are also checked. 

Another nice thing about the discovery mechanics is that if, for instance wants to make several versions of a compiled simulator (e.g. different optimizations) it would be easier to implement. 

We could also use the discovery function to "discover" online backends and make them subclass BaseBackend. This would make us less dependent on changes to the set of available online backends.

I'm a little unclear about what niceties we are giving up. In the previous discovery I intentionally hid backends and just made LocalSimulator the only one that was visible. With this PR I import all the backend modules into the namespace as you suggested so you should see no difference from doing a normal import (that was the goal anyway). For instance documentation and tab completion seems to be available now.